src/Compilers/Core/Portable/Operations/OperationExtensions.cs

```diff
@@ -1,13 +1,76 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Threading;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Semantics
 {
     public static class OperationExtensions
     {
+        /// <summary>
+        /// Find the argument supplied for a given parameter of the target method.
+        /// </summary>
+        /// <param name="hasArgumentsExpression">The IHasArgumentsExpression object to get matching argument object from.</param>
+        /// <param name="parameter">Parameter of the target method.</param>
+        /// <returns>Argument corresponding to the parameter.</returns>
+        public static IArgument GetArgumentMatchingParameter(this IHasArgumentsExpression hasArgumentsExpression, IParameterSymbol parameter)
+        {
+            if (hasArgumentsExpression == null || parameter == null)
+            {
+                return null;
+            }
+
+            foreach (var argument in hasArgumentsExpression.ArgumentsInEvaluationOrder)
+            {
+                if (argument.Parameter == parameter)
+                {
+                    return argument;
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Arguments of the invocation, excluding the instance argument. Arguments are in parameter order,
+        /// and params/ParamArray arguments have been collected into arrays. Default values are supplied for
+        /// optional arguments missing in source.
+        /// </summary>
+        public static ImmutableArray<IArgument> ArgumentsInParameterOrder(this IHasArgumentsExpression hasArgumentsExpression)
+        {
+            ImmutableArray<IArgument> argumentsInEvaluationOrder = hasArgumentsExpression.ArgumentsInEvaluationOrder;
+
+            // If all of the arguments were specified positionally the evaluation order is the same as the parameter order.
+            if (ArgumentsAreInParameterOrder(argumentsInEvaluationOrder))
+            {
+                return argumentsInEvaluationOrder;
+            }
+
+            return argumentsInEvaluationOrder.Sort(
+                (x, y) =>
+                {
+                    int x1 = x.Parameter?.Ordinal ?? -1;
```

Comment:
>x.Parameter?.Ordinal ?? -1 [](start = 29, length = 26)

I am not sure if this approach guarantees deterministic result if there are more than one argument without parameter. Also, the contract of the method doesn't say what should happen with such arguments. I think they shouldn't be included into the result at all. It appears that the current implementation places them at the beginning of the result, making it impossible to reliably index into the result by using parameter's ordinal.